File: 139_Word_Break.py
class Solution:
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        dp = [False] * len(s)
        for i in range(len(s)):
            for w in wordDict:
                if (i - len(w) == -1 or dp[i - len(w)]) and w == s[i - len(w) + 1:i + 1]:
                    dp[i] = True
                    break
        return dp[-1]

File: 139_Word_Break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp (s.size() + 1, false);
        dp[0] = true;
        for (int i = 0; i < s.size(); ++i) {
            for (string word : wordDict) {
                int n = word.size();
                if ((i - n >= -1 && dp[i - n + 1]) && s.substr(i - n + 1, n) == word) {
                    dp[i + 1] = true;
                }
            }
        }
        return dp[s.size()];
    }
};
